accumulator/accumulate_signals_workflow.go

```diff
@@ -0,0 +1,168 @@
+	BucketKey          string
+	GreetingsList      []AccumulateGreeting
+	UniqueGreetingKeys map[string]bool
+	startTime          time.Time
+}
+
+// GetNextTimeout returns the maximum time for a workflow to wait for the next signal.
+// This waits for the greater of the remaining fromStartTimeout and  signalToSignalTimeout
+// fromStartTimeout and signalToSignalTimeout can be adjusted to wait for the right amount of time as desired
+// This resets with Continue As New
+func (a *AccumulateGreeting) GetNextTimeout(ctx workflow.Context, startTime time.Time, exitRequested bool) (time.Duration, error) {
+	if exitRequested {
+		return exitTimeout, nil
+	}
+	if startTime.IsZero() {
+		startTime = workflow.GetInfo(ctx).WorkflowStartTime // if you want to start from the time of the first signal, customize this
+	}
+	total := workflow.Now(ctx).Sub(startTime)
+	totalLeft := fromStartTimeout - total
+	if totalLeft <= 0 {
+		return 0, nil
+	}
+	if signalToSignalTimeout > totalLeft {
+		return signalToSignalTimeout, nil
+	}
+	return totalLeft, nil
+}
+
+// AccumulateSignalsWorkflow workflow definition
+func AccumulateSignalsWorkflow(ctx workflow.Context, greetings GreetingsInfo) (allGreetings string, err error) {
+	log := workflow.GetLogger(ctx)
+	var a AccumulateGreeting
+	if greetings.GreetingsList == nil {
+		greetings.GreetingsList = []AccumulateGreeting{}
+	}
+	if greetings.UniqueGreetingKeys == nil {
+		greetings.UniqueGreetingKeys = make(map[string]bool)
+	}
+	var unprocessedGreetings []AccumulateGreeting
+	if greetings.startTime.IsZero() {
+		greetings.startTime = workflow.Now(ctx)
+	}
+	exitRequested := false
+
+	ao := workflow.ActivityOptions{
+		StartToCloseTimeout: 100 * time.Second,
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+
+	for !workflow.GetInfo(ctx).GetContinueAsNewSuggested() {
+
+		timeout, err := a.GetNextTimeout(ctx, greetings.startTime, exitRequested)
+		childCtx, cancelHandler := workflow.WithCancel(ctx)
+		selector := workflow.NewSelector(ctx)
+
+		if err != nil {
+			log.Error("Error calculating timeout")
+			return "", err
+		}
+		log.Debug("Awaiting for " + timeout.String())
+		selector.AddReceive(workflow.GetSignalChannel(ctx, "greeting"), func(c workflow.ReceiveChannel, more bool) {
+			c.Receive(ctx, &a)
+			unprocessedGreetings = append(unprocessedGreetings, a)
+			log.Debug("Signal Received with text: " + a.GreetingText + ", more?: " + strconv.FormatBool(more) + "\n")
+			cancelHandler() // cancel timer future
+			a = AccumulateGreeting{}
+		})
+		selector.AddReceive(workflow.GetSignalChannel(ctx, "exit"), func(c workflow.ReceiveChannel, more bool) {
+			c.Receive(ctx, nil)
+			exitRequested = true
+			cancelHandler() // cancel timer future
+			log.Debug("Exit Signal Received, more?: " + strconv.FormatBool(more) + "\n")
+		})
+
+		timerFuture := workflow.NewTimer(childCtx, timeout)
+		selector.AddFuture(timerFuture, func(f workflow.Future) {
+			log.Debug("Timer fired \n")
+		})
+
+		selector.Select(ctx)
```

Comment:
Can you help me understand the goal here? Are you wanting to wait for the first signal or timer fired? Or do you want to collect all signals until timer reached and then process them? You need to loop on the select until timer fired if you want the latter and not cancel the timer on each signal.